test/infrastructure/docker/internal/controllers/devmachine_controller.go

```diff
@@ -0,0 +1,257 @@
+// +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;machinesets;machines,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=secrets;,verbs=get;list;watch
+// +kubebuilder:rbac:groups="",resources=configmaps,verbs=get;list;watch
+
+// Reconcile handles DevMachine events.
+func (r *DevMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, rerr error) {
+	ctx = container.RuntimeInto(ctx, r.ContainerRuntime)
+
+	// Fetch the DevMachine instance.
+	devMachine := &infrav1.DevMachine{}
+	if err := r.Client.Get(ctx, req.NamespacedName, devMachine); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	if finalizerAdded, err := finalizers.EnsureFinalizer(ctx, r.Client, devMachine, infrav1.MachineFinalizer); err != nil || finalizerAdded {
+		return ctrl.Result{}, err
+	}
+
+	// AddOwners adds the owners of DevMachine as k/v pairs to the logger.
+	// Specifically, it will add KubeadmControlPlane, MachineSet and MachineDeployment.
+	ctx, log, err := clog.AddOwners(ctx, r.Client, devMachine)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Fetch the Machine.
+	machine, err := util.GetOwnerMachine(ctx, r.Client, devMachine.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if machine == nil {
+		log.Info("Waiting for Machine Controller to set OwnerRef on DevMachine")
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("Machine", klog.KObj(machine))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("DevMachine owner Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, err
+	}
+	if cluster == nil {
+		log.Info(fmt.Sprintf("Please associate this machine with a cluster using the label %s: <name of cluster>", clusterv1.ClusterNameLabel))
+		return ctrl.Result{}, nil
+	}
+
+	log = log.WithValues("Cluster", klog.KObj(cluster))
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	if isPaused, conditionChanged, err := paused.EnsurePausedCondition(ctx, r.Client, cluster, devMachine); err != nil || isPaused || conditionChanged {
+		return ctrl.Result{}, err
+	}
+
+	if cluster.Spec.InfrastructureRef == nil {
+		log.Info("Cluster infrastructureRef is not available yet")
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch the DevCluster.
+	devCluster := &infrav1.DevCluster{}
+	devClusterName := client.ObjectKey{
+		Namespace: devMachine.Namespace,
+		Name:      cluster.Spec.InfrastructureRef.Name,
+	}
+	if err := r.Client.Get(ctx, devClusterName, devCluster); err != nil {
+		log.Info("DockerCluster is not available yet")
+		return ctrl.Result{}, nil
+	}
+
+	backendReconciler := r.backendReconcilerFactory(ctx, devMachine)
```

Comment:
Same here. Let's please do this in SetupWithManager

Reply:
For DevMachine we can't, otherwise we can't support DevMachine with different backends.